Document managing users, roles and accesslists through IaC (#36622)
gravitational/teleport: docs/pages/management/dynamic-resources/user-and-role.mdx

---

> title: Managing users and roles with IaC

Reviewer: This should be title case

> description: Use Infrastructure-as-Code tooling to create Teleport users and roles.

Reviewer: ```suggestion
description: Use infrastructure-as-code tooling to create Teleport users and roles.
```
I tend to see this uncapitalized in body text

---

In this guide, you will see how to create users and grant them roles through

> Infrastructure-as-Code (IaC). Teleports supports three ways to dynamically

Reviewer: ```suggestion
infrastructure as code (IaC). Teleports supports three ways to dynamically
```

create resources from code:

> - The Teleport Kubernetes Operator that allows you to manage Teleport resources
>   from Kubernetes
> - The Teleport Terraform Provider that allows you to manage Teleport Resources
>   via Terraform
> - The `tctl` CLI that allows you to manage Teleport resources from your local
>   computer or your CI environment
> 

Reviewer: ```suggestion
- The Teleport Kubernetes Operator, which allows you to manage Teleport resources
  from Kubernetes
- The Teleport Terraform Provider, which allows you to manage Teleport resources
  via Terraform
- The `tctl` CLI, which allows you to manage Teleport resources from your local
  computer or your CI environment

```

> ### Prerequisites

Reviewer: The first-level heading should be H2. We should bump all headings in the page down a level.

To follow this guide, you must have:

<Tabs>
<TabItem label="tctl">

> (!docs/pages/includes/edition-prereqs-tabs.mdx!)

Reviewer: Instead of this partial, I would include instructions to download `tctl`.

Also, if we are going to include adding a Teleport cluster as a prerequisite, we should add it for tctl, the Operator, and the Terraform provider.

Author: For the operator and the terraform cluster, the Teleport cluster is a requirement in their respective guides.
If you have a working operator/TF provider, you have a running Teleport cluster somewhere.

The operator and terraform requirements are a bit specific wrt to the kind of teleport cluster they can connect to, so I think it's better to leave to teleport requirement in the dedicated guide instead of putting an awkward generic requirement here.

</TabItem>

> <TabItem label="operator">

Reviewer: ```suggestion
<TabItem label="Kubernetes Operator">
```

A running operator by following either:
- [the guide to enable the operator in the `teleport-cluster` Helm chart](./teleport-operator-helm.mdx).
- [the guide to setup a standalone operator](./teleport-operator-standalone.mdx).

You must also set the namespace in which you deployed the operator as this is
the namespace where you will deploy the CustomResources:

```code
# for operators deployed with the `teleport-cluster` Helm chart
$ export OPERATOR_NAMESPACE="teleport-cluster"

# for standalone operators
$ export OPERATOR_NAMESPACE="teleport-iac"
```

</TabItem>
<TabItem label="terraform">

A functional Teleport Terraform provider by following [the Terraform provider guide](./terraform-provider.mdx).

</TabItem>
</Tabs>

> ### Step 1/4 - Write manifests

Reviewer: "Step" sections use the "Step n/d. " format. (This applies to other "Step" sections on this page.)

In this step, we'll write text files describing the resources we want in
Teleport. Those files are called manifests and their syntax will vary based on
the IaC tooling you'll use.

Those manifests are typically versioned in a shared revision system like git.
This allows you to keep track of all changes, follow standard code review
procedures before changing resources in Teleport, and quickly redeploy your
Teleport instance if needed.

#### Write role manifests

We will create 2 roles:

> - `manager` that allows listing users, roles and reviewing audit events and
>   session contents.
> - `engineer` that grants access to dev and staging servers.

Reviewer: ```suggestion
- `manager` allows listing users and roles, as well as reviewing audit events and
  session contents.
- `engineer` grants access to dev and staging servers.
```
grammar fixes

$ kubectl apply -n "$OPERATOR_NAMESPACE" -f users.yaml 
teleportuser.resources.teleport.dev/alice created
teleportuser.resources.teleport.dev/bob created
```

List the created Kubernetes resources:

```code
$ kubectl get teleportroles -n "$OPERATOR_NAMESPACE"
# NAME       AGE
# engineer   10m
# manager    10m

$ kubectl get teleportusers -n "$OPERATOR_NAMESPACE"
# NAME     AGE
# alice    10m
# bob      10m
```

</TabItem>
<TabItem label="terraform">

```code
$ terraform plan
[...]
Plan: 4 to add, 0 to change, 0 to destroy.

$ terraform apply
teleport_role.engineer: Creating...
teleport_role.manager: Creating...
teleport_role.engineer: Creation complete after 0s [id=engineer]
teleport_role.manager: Creation complete after 0s [id=manager]
teleport_user.bob: Creating...
teleport_user.alice: Creating...
teleport_user.bob: Creation complete after 0s [id=bob]
teleport_user.alice: Creation complete after 0s [id=alice]

Apply complete! Resources: 4 added, 0 changed, 0 destroyed.
```

</TabItem>
</Tabs>

### Step 3/4 - Validate users were created

Now that the IaC tooling has run, we'll validate that the users were properly
created and granted the correct roles.

<Tabs>

> <TabItem label="via the UI">

Reviewer: ```suggestion
<TabItem label="Via the UI">
```

If you have UI access, connect to your Teleport cluster Web UI, open the
management panel, and select the "Users" tab.

![Screenshot of the web UI listing alice and bob users](../../../img/management/check-users-web-ui.png)

Two new users `alice` and `bob` should be present.

</TabItem>

> <TabItem label="via CLI">

Reviewer: ```suggestion
<TabItem label="Via CLI">
```

----------------------------- -------------------------
@teleport-access-approval-bot @teleport-access-approver
alice                         manager,engineer
bob                           engineer
bot-operator                  bot-operator

# alternatively you can inspect the users details by doing
$ tctl get user/alice
kind: user
metadata:
  id: 1704849160091933780
  labels:
    teleport.dev/origin: kubernetes
  name: alice
spec:
  created_by:
    time: "2024-01-10T01:12:40.088581806Z"
    user:
      name: bot-operator
  expires: "0001-01-01T00:00:00Z"
  roles:
  - manager
  - engineer
  status:
    is_locked: false
    lock_expires: "0001-01-01T00:00:00Z"
    locked_time: "0001-01-01T00:00:00Z"
    recovery_attempt_lock_expires: "0001-01-01T00:00:00Z"
version: v2
```

</TabItem>
</Tabs>

### Step 4/4 - Create a password reset link

At this point, the local users have been created in Teleport. However, we never
specified any password or additional authentications factors. You must issue a
password reset link for the users to finish their Teleport registration and be
able to log in Teleport.

User reset links contain single-use expiring tokens. Because of this, you cannot

> follow the same descriptive approach as for other Teleport resources and

Reviewer: Should "descriptive" be "declarative" here?

generate them via a manifest. You need to create those tokens once after the
user creation, and securely send them to the end-user for them to register their
password/MFA.

> <Tabs>

Reviewer: I would remove the `<Tab(Item)>` tags and incorporate all of these instructions into the body text so users will see both options. This is because a user may not necessarily know whether they want to use `tctl` or build automation until they have read all of the instructions here.

<TabItem label="user reset via CLI">
You can manually reset a user password via `tctl` by doing:

```code
$ tctl users reset alice
User "alice" has been reset. Share this URL with the user to complete password reset, link is valid for 8h:
https://teleport.example.com:443/web/reset/05b420fdc784597cbbb1d2ba65697cd8

NOTE: Make sure teleport.example.com:443 points at a Teleport proxy which users can access.
```

</TabItem>
<TabItem label="automating user reset">

If you have a way to securely send reset links to the users, you can build
automation to fit your organization's specific needs. For example:

```
$ tctl users reset alice --format=json | \
    jq '"Sending an email to " + .spec.user +" that contains the link: " + .spec.url'
```

You must replace the `jq` command by something that actually sends the link over

> a secure channel. This channel will depend on your organization, it is usually a
> direct message, or an email.

Reviewer: ```suggestion
a secure channel. This channel will depend on your organization. It is usually a
direct message or an email.
```

<Admonition type="tip" title="For Terraform users">

You can trigger your custom script on Terraform resource creation with
[the `local-exec` provisioner](https://developer.hashicorp.com/terraform/language/resources/provisioners/local-exec).

```hcl
resource "teleport_user" "bob" {
  version = v2
  metadata = {
    name = "bob"
  }

  spec = {
    roles = [teleport_role.engineer.metadata.name]
  }
  
  # on user creation, trigger a reset flow and send the link via 
  provisioner "local-exec" {
    command = "tctl users reset alice --format=json | jq '\"Sending an email to \" + .spec.user +\" that contains the link: \" + .spec.url'"
  }
}

```

</Admonition>
</TabItem>
</Tabs>

> ### Next Steps

Reviewer: ```suggestion
### Next steps
```
We use sentence case in section headings

- Allow users with the `manager` role to grant access to production servers to
  some `engineers` via Access Lists. Manager will need to justify and review
  granted access periodically.
  See [the AccessList documentation](../../access-controls/access-lists.mdx) for
  a high-level explanation of the feature,
  and [the AccessList IaC guide](access-list.mdx) for a step by step IaC
  AccessList setup.
- Allow users with the `engineer` role to request temporary access to
  production, and have users with the `manager` role validate the requests.

>   See [the access-requests documentation](../../access-controls/access-requests.mdx)

Reviewer: ```suggestion
  See [the Access Requests documentation](../../access-controls/access-requests.mdx)
```
